File: data-jpa/src/main/java/org/appsugar/integration/data/jpa/entity/IdEntity.java
package org.appsugar.integration.data.jpa.entity;

import lombok.Getter;
import lombok.Setter;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.GenericGenerator;
import org.hibernate.annotations.UpdateTimestamp;

import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.MappedSuperclass;
import java.util.Date;
import java.util.Objects;

/**
 * 基础对象类
 * 抽象化提供kotlin进行继承,方便data class 包含
 *
 * @author shenliuyang
 * @version 1.0.0
 * @package org.appsugar.integration.data.jpa
 * @className IdEntity
 * @date 2021-04-08  19:13
 */
@MappedSuperclass
@Getter
@Setter
public abstract class IdEntity {
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO, generator = "native")
    @GenericGenerator(name = "native", strategy = "native")
    protected Long id;
    @CreationTimestamp
    protected Date createdAt;
    @UpdateTimestamp
    protected Date updatedAt;


    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        IdEntity that = (IdEntity) o;

        return Objects.equals(getId(), that.getId());
    }

    @Override
    public int hashCode() {
        return getId() != null ? getId().hashCode() : super.hashCode();
    }


}

File: data-jpa/src/main/kotlin/org/appsugar/integration/data/jpa/entity/IdEntity.kt
package org.appsugar.integration.data.jpa.entity

import org.hibernate.annotations.CreationTimestamp
import org.hibernate.annotations.GenericGenerator
import org.hibernate.annotations.UpdateTimestamp
import java.io.Serializable
import java.util.*
import javax.persistence.GeneratedValue
import javax.persistence.GenerationType
import javax.persistence.Id
import javax.persistence.MappedSuperclass

@MappedSuperclass
abstract class IdEntity : Serializable {
    @get:Id
    @get:GeneratedValue(strategy = GenerationType.AUTO, generator = "native")
    @get:GenericGenerator(name = "native", strategy = "native")
    abstract var id: Long

    @get:CreationTimestamp
    abstract var createdAt: Date?

    @get:UpdateTimestamp
    abstract var updatedAt: Date?
}